### PR TITLE
build: record git sha so that we can write it into binaries

### PR DESCRIPTION
**Pull request by tylershunt**
Having cmake config depend on `.git/index` ensures that the current
sha will be recorded whenever git state changes.

Also a fallback to record "unknown" if for some weird reason git is not
installed.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

### CODE REVIEW

File: cmake/Modules/GitHeadSHA.cmake [hunk 1]
```diff
@@ -0,0 +1,17 @@
+find_package(Git QUIET)
+
+if (Git_FOUND)
+execute_process(
```
Comment (ddn0) [inline cmake/Modules/GitHeadSHA.cmake:4]: can you append something to indicate if the build is from a clean or dirty working directory?

Comment (tylershunt) [reply cmake/Modules/GitHeadSHA.cmake:4]: It's more complicated to get the working directory state without forcing a re-configure on every build. I'm working that up in my free time.

Comment (ddn0) [reply cmake/Modules/GitHeadSHA.cmake:4]: I don't think it's worth complicating the build to get a "dirty" flag in.

We probably can get the same value by doing the complement and just having a way to specially mark release builds (e.g., CI conda builds) FWIW.

I'm okay with merging as is.

**Comment by tylershunt** [comment]
I added the build type (e.g. `f6c9d19a9-release`) because that seemed like a good idea too.

**Review by ddn0 — APPROVED** [review]

**Pull request merged**